dnfdaemon-client/commands/command.cpp

```
@@ -0,0 +1,78 @@
+
+Dnfdaemon-client is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with dnfdaemon-client.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "command.hpp"
+
+#include "../context.hpp"
+#include "../utils.hpp"
+
+#include <dnfdaemon-server/dbus.hpp>
+#include <libdnf/transaction/transaction_item_action.hpp>
+
+#include <iostream>
+#include <vector>
+
+namespace dnfdaemon::client {
+
+void TransactionCommand::print_transaction(std::vector<dnfdaemon::DbusTransactionItem> transaction) {
+    // TODO(mblaha): print decent transaction table using smartcols
+    for (auto & ti : transaction) {
+        libdnf::transaction::TransactionItemAction action =
+            static_cast<libdnf::transaction::TransactionItemAction>(std::get<0>(ti));
+        std::cout << "Action: " << libdnf::transaction::TransactionItemAction_get_name(action) << std::endl;
+        std::cout << "Name: " << std::get<1>(ti) << std::endl;
+        std::cout << "Arch: " << std::get<5>(ti) << std::endl;
+        std::cout << "EVR: " << std::get<2>(ti) << ":" << std::get<3>(ti) << "-" << std::get<4>(ti) << std::endl;
+        std::cout << "Repo: " << std::get<6>(ti) << std::endl;
+    }
+}
+
+void TransactionCommand::run_transaction(Context & ctx) {
+    dnfdaemon::KeyValueMap options = {};
+
+    // resolve the transaction
+    options["allow_erasing"] = ctx.allow_erasing->get_value();
+    std::vector<dnfdaemon::DbusTransactionItem> transaction;
+    ctx.session_proxy->callMethod("resolve")
+        .onInterface(dnfdaemon::INTERFACE_RPM)
+        .withTimeout(-1)
+        .withArguments(options)
+        .storeResultsTo(transaction);
+    if (transaction.empty()) {
+        std::cout << "Nothing to do." << std::endl;
+        return;
+    }
```

How do you suppose to handle an arror state? I suggest that in this case the transaction is also empty?

---

In case of "hard" failure (e.g. --strict is used and package is not installable) the "resolve" method call will return dbus error reply. The sdbus-cpp library then throws it as an exception with error description like
```
terminate called after throwing an instance of 'sdbus::Error'
  what():  [org.rpm.dnf.v0.Error] [org.rpm.dnf.v0.rpm.Rpm.ResolveError] Problem: conflicting requests
  - nothing provides XXXNot needed by Failing-1.0.1-0.x86_64
```
Unfortunately the exception is at the moment silently suppressed in `main()` function - I will fix it and print the error message to the user. This is the easy part.

But you are right that there might occur non fatal resolving errors that and are currently not returned by `resolve` call. I will add an output parameter containing a list of problems. Then client could print those "warnings" before actual transaction table.